### PR TITLE
fix remove unique constraint for collection_series.

### LINKED ISSUE #986: /collection/{slug}: exception happens when user adds a series to collection that already has this series
```
018-12-19 00:13:54.532 ERROR 2181 --- [tp2100942741-24] r.m.web.controller.ErrorController       : Exception 'org.springframework.web.util.NestedServletException' occurred on page /series/1

org.springframework.web.util.NestedServletException: Request processing failed; nested exception is org.springframework.dao.DuplicateKeyException: PreparedStatementCallback; SQL [INSERT INTO collections_series ( collection_id , series_id , number_of_stamps , price , currency ) SELECT c.id AS collection_id , ? AS series_id , ? AS number_of_stamps , ? as price , ? as currency FROM collections c WHERE c.user_id = ?]; Нарушение уникального индекса или первичного ключа: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"
Unique index or primary key violation: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"; SQL statement:
INSERT INTO collections_series ( collection_id , series_id , number_of_stamps , price , currency ) SELECT c.id AS collection_id , ? AS series_id , ? AS number_of_stamps , ? as price , ? as currency FROM collections c WHERE c.user_id = ? [23505-197]; nested exception is org.h2.jdbc.JdbcSQLException: Нарушение уникального индекса или первичного ключа: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"
Unique index or primary key violation: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"; SQL statement:
INSERT INTO collections_series ( collection_id , series_id , number_of_stamps , price , currency ) SELECT c.id AS collection_id , ? AS series_id , ? AS number_of_stamps , ? as price , ? as currency FROM collections c WHERE c.user_id = ? [23505-197]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:982) ~[spring-webmvc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:872) ~[spring-webmvc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707) ~[javax.servlet-api-3.1.0.jar:3.1.0]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846) ~[spring-webmvc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790) ~[javax.servlet-api-3.1.0.jar:3.1.0]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812) ~[jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669) ~[jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:124) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:66) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177) ~[spring-security-web-4.2.9.RELEASE.jar:4.2.9.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:347) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:263) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:109) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652) ~[jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585) [jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577) ~[jetty-security-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223) ~[jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127) ~[jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515) [jetty-servlet-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) ~[jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061) ~[jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.servlets.gzip.GzipHandler.handle(GzipHandler.java:529) [jetty-servlets-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97) [jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.Server.handle(Server.java:499) [jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311) [jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257) [jetty-server-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544) [jetty-io-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635) [jetty-util-9.2.18.v20160721.jar:9.2.18.v20160721]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555) [jetty-util-9.2.18.v20160721.jar:9.2.18.v20160721]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
Caused by: org.springframework.dao.DuplicateKeyException: PreparedStatementCallback; SQL [INSERT INTO collections_series ( collection_id , series_id , number_of_stamps , price , currency ) SELECT c.id AS collection_id , ? AS series_id , ? AS number_of_stamps , ? as price , ? as currency FROM collections c WHERE c.user_id = ?]; Нарушение уникального индекса или первичного ключа: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"
Unique index or primary key violation: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"; SQL statement:
INSERT INTO collections_series ( collection_id , series_id , number_of_stamps , price , currency ) SELECT c.id AS collection_id , ? AS series_id , ? AS number_of_stamps , ? as price , ? as currency FROM collections c WHERE c.user_id = ? [23505-197]; nested exception is org.h2.jdbc.JdbcSQLException: Нарушение уникального индекса или первичного ключа: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"
Unique index or primary key violation: "UC_COLLECTIONS_SERIES_COLLECTION_ID_SERIES_ID_INDEX_1 ON PUBLIC.COLLECTIONS_SERIES(COLLECTION_ID, SERIES_ID) VALUES (1, 1, 1)"; SQL statement:
INSERT INTO collections_series ( collection_id , series_id , number_of_stamps , price , currency ) SELECT c.id AS collection_id , ? AS series_id , ? AS number_of_stamps , ? as price , ? as currency FROM collections c WHERE c.user_id = ? [23505-197]
	at org.springframework.jdbc.support.SQLErrorCodeSQLExceptionTranslator.doTranslate(SQLErrorCodeSQLExceptionTranslator.java:238) ~[spring-jdbc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:73) ~[spring-jdbc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:655) ~[spring-jdbc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.update(JdbcTemplate.java:876) ~[spring-jdbc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.jdbc.core.JdbcTemplate.update(JdbcTemplate.java:900) ~[spring-jdbc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate.update(NamedParameterJdbcTemplate.java:287) ~[spring-jdbc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate.update(NamedParameterJdbcTemplate.java:292) ~[spring-jdbc-4.3.20.RELEASE.jar:4.3.20.RELEASE]
	at ru.mystamps.web.feature.collection.JdbcCollectionDao.addSeriesToUserCollection(JdbcCollectionDao.java:232) ~[classes/:na]
	at ru.mystamps.web.feature.collection.CollectionServiceImpl.addToCollection(CollectionServiceImpl.java:90) ~[classes/:na]
...
```

Solution:
- we should remove unique constraint from collection table to allow to have more than one instance of a series

### PR DESCRIPTION
**Pull request by mukeshk**
Fix #986

### CODE REVIEW

**Comment by mukeshk** [comment]
On my local pc,

alter table `mystamps`.`collections_series` drop index `uc_collections_series_collection_id_series_id`

Error on rename of '.\mystamps\#sql-c38_6' to '.\mystamps\collections_series' (errno: 150 - Foreign key constraint is incorrectly formed)

**Comment by mukeshk** [comment]
image

**Comment by mukeshk** [comment]
Alter table `mystamps`.`collections_series`   
  drop index `uc_collections_series_collection_id_series_id`,
  drop foreign key `fk_collections_series_collection_id`,
  drop foreign key `fk_collections_series_series_id`

MysqlError #1025
Error on rename of '.\mystamps\collections_series' to '.\mystamps\#sql2-c38-6' (errno: 152 - Cannot delete a parent row)

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd">
+		
+	<changeSet author="mukeshk" id="drop-unique-constraint-from-collections_series-table">
+		<comment>Remove unique constraint (collection_id,series_id) from collection_series</comment>
+		<dropForeignKeyConstraint baseTableName="collections_series"  constraintName="fk_collections_series_series_id" />
```
Comment (php-coder) [inline src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: It's counter intuitive that we remove and re-create foreign keys. Is it required? Would `dropUniqueConstraint` work without it?

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd">
+		
+	<changeSet author="mukeshk" id="drop-unique-constraint-from-collections_series-table">
```
Comment (php-coder) [inline src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: Typically "id" come first and also "context" attribute is missing here.

**Comment by php-coder** [comment]
Ok, I'll look on this failure later.

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml (hunk 1 above)
Comment (php-coder) [reply src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: This line has trailing space BTW.

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+		<comment>Remove unique constraint (collection_id,series_id) from collection_series</comment>
+		<dropForeignKeyConstraint baseTableName="collections_series"  constraintName="fk_collections_series_series_id" />
+		<dropUniqueConstraint constraintName="uc_collections_series_collection_id_series_id"  tableName="collections_series"/>
```
Comment (php-coder) [inline src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: This line has trailing space BTW.

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd">
+		
+	<changeSet author="mukeshk" id="drop-unique-constraint-from-collections_series-table">
+		<comment>Remove unique constraint (collection_id,series_id) from collection_series</comment>
+		<dropForeignKeyConstraint baseTableName="collections_series"  constraintName="fk_collections_series_series_id" />
+		<dropUniqueConstraint constraintName="uc_collections_series_collection_id_series_id"  tableName="collections_series"/>	
+		<addForeignKeyConstraint baseTableName="collections_series"  baseColumnNames="series_id"
```
Comment (php-coder) [inline src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: This line has trailing space BTW.

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd">
+		
+	<changeSet author="mukeshk" id="drop-unique-constraint-from-collections_series-table">
+		<comment>Remove unique constraint (collection_id,series_id) from collection_series</comment>
```
Comment (php-coder) [inline src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: I'd rather remove this comment as it looks obvious to me.

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml (hunk 1 above)
Comment (php-coder) [reply src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: > Would dropUniqueConstraint work without it?

I've just tested it locally (with H2) and it works. So, it's not clear so far why you've added it as it's unnecessary.

**Comment by mukeshk** [comment]
Yes. The drop unique constraint worked on H2 db. But Travis failed for MySQL

I will push again so that you will notice in Travis.

Thanks

**Comment by php-coder** [comment]
> Yes. The drop unique constraint worked on H2 db. But Travis failed for MySQL

I'll look on this but anyway it should be noted in the comments. Mostly anything that we solve in a unexpected way because of the obstacles, should be expressed in the comments.

> I will push again so that you will notice in Travis.

JFYI: I don't get notified on your builds. So the only way to see is when you push to this PR and I look on the result manualy.

File: src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd">
+		
+	<changeSet author="mukeshk" id="drop-unique-constraint-from-collections_series-table">
+		<dropUniqueConstraint constraintName="uc_collections_series_collection_id_series_id"  tableName="collections_series"/>
```
Comment (php-coder) [inline src/main/resources/liquibase/version/0.4.1/2019-09-18--rollback_collection_unique_series_constraint.xml]: There are 2 spaces before "tableName" attribute :)

**Comment by php-coder** [comment]
I'm a little confused why it uses `DROP KEY` instead of `DROP INDEX`:

> Caused by: liquibase.exception.DatabaseException: Error on rename of './mystamps/#sql-e87_6' to './mystamps/collections_series' (errno: 150 - Foreign key constraint is incorrectly formed) [Failed SQL: ALTER TABLE mystamps.collections_series DROP KEY uc_collections_series_collection_id_series_id]

I wouldn't surprise if that's a liquibase bug, for instance.

**Comment by php-coder** [comment]
> I'm a little confused why it uses DROP KEY instead of DROP INDEX:

Both are failing in this case :(

```
mysql> alter table collections_series2 drop key uc_collections_series_collection_id_series_id;
ERROR 1025 (HY000): Error on rename of './mystamps/#sql-488_b7' to './mystamps/collections_series2' (errno: 150 - Foreign key constraint is incorrectly formed)
mysql> alter table collections_series2 drop index uc_collections_series_collection_id_series_id;
ERROR 1025 (HY000): Error on rename of './mystamps/#sql-488_b7' to './mystamps/collections_series2' (errno: 150 - Foreign key constraint is incorrectly formed)
```

**Comment by php-coder** [comment]
I found an advise (see https://dba.stackexchange.com/questions/190153/mysql-create-table-error-1005-errno-150-foreign-key-constraint-is-incorrectly/201950#201950) to use `SHOW ENGINE INNODB STATUS`. Here is the related output:

```
------------------------
LATEST FOREIGN KEY ERROR
------------------------
2019-09-21 20:14:56 0x7fbeb021d700 Error in foreign key constraint of table mystamps/collections_series2:
there is no index in the table which would contain
the columns as the first columns, or the data types in the
table do not match the ones in the referenced table
or one of the ON ... SET NULL columns is declared NOT NULL. Constraint:
,
  CONSTRAINT `collections_series2_ibfk_1` FOREIGN KEY (`collection_id`) REFERENCES `collections` (`id`)
```
After closer look on the table definition I see the following:
```
mysql> show create table collections_series2;
...
  UNIQUE KEY `uc_collections_series_collection_id_series_id` (`collection_id`,`series_id`),
  KEY `fk_collections_series_series_id` (`series_id`),
  CONSTRAINT `collections_series2_ibfk_1` FOREIGN KEY (`collection_id`) REFERENCES `collections` (`id`),
  CONSTRAINT `collections_series2_ibfk_2` FOREIGN KEY (`series_id`) REFERENCES `series`
...
```
Note that we have `KEY `fk_collections_series_series_id` (`series_id`)` that creates an index for `series_id` field but there is no similar key for `collection_id` field. Why? I suppose this is because MySQL uses compound index instead (uc_collections_series_collection_id_series_id). And when we try to remove it, we get refused because removal leads to the state when this field won't have an index at all. And we have to have an index because this field has a foreign key. So, let's try to add this index manually:
```
mysql> alter table collections_series2 add key fk_collections_series_collection_id(collection_id);
Query OK, 0 rows affected (0.01 sec)
Records: 0  Duplicates: 0  Warnings: 0
mysql> alter table collections_series2 drop key uc_collections_series_collection_id_series_id;
Query OK, 0 rows affected (0.02 sec)
Records: 0  Duplicates: 0  Warnings: 0
mysql> show create table collections_series2;
...
  KEY `fk_collections_series_series_id` (`series_id`),
  KEY `fk_collections_series_collection_id` (`collection_id`),
  CONSTRAINT `collections_series2_ibfk_1` FOREIGN KEY (`collection_id`) REFERENCES `collections` (`id`),
  CONSTRAINT `collections_series2_ibfk_2` FOREIGN KEY (`series_id`) REFERENCES `series` (`id`)
...
```
So, it seems like we should add this key then before dropping an unique constraint.

I'll check how it works in H2 later to confirm whether it's right way to go.

**Comment by php-coder** [comment]
> I'll check how it works in H2 later to confirm whether it's right way to go.

It looks like H2 has 2 indexes already and it was MySQL trick to optimize number of indices. So, I suggest to: before dropping a unique index, first create an index for collection_id column. It will be a separate changeset I think. If this operation fails on H2, then we should execute it only on MySQL (with `dbms="mysql"` attribute).

**Comment by mukeshk** [comment]
It seems I messed up with my branch. it created a detached head.
> git status
HEAD detached from 08d659f6
nothing to commit, working tree clean

> git rebase master origin/gh986_rollback_collection_series_unique_constraint -i
Successfully rebased and updated detached HEAD.

> git status
On branch gh986_rollback_collection_series_unique_constraint
Your branch is up to date with 'origin/gh986_rollback_collection_series_unique_constraint'.

> git rebase master origin/gh986_rollback_collection_series_unique_constraint -i
Successfully rebased and updated detached HEAD.

> git checkout master
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 4 commits.
  (use "git push" to publish your local commits)

Any idea what i can do to fix this? sorry for the trouble.

**Comment by php-coder** [comment]
> Your branch is ahead of 'origin/master' by 4 commits.

If you only need to reset master to be even with origin/master, use `git reset --hard origin/master` within `master` branch. This removes 4 commits. If you need them, copy them with `git cherry-pick` to a branch where you need them.

File: src/main/resources/liquibase/version/0.4.1/2019-09-29--create_index_series_id_on_collection_series.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.0.xsd">
+		
+	<changeSet author="mukeshk" id="create-index-seriesid-on-collections-series-table">
```
Comment (php-coder) [inline src/main/resources/liquibase/version/0.4.1/2019-09-29--create_index_series_id_on_collection_series.xml]: Let's put 2 change sets to a single file.

**Comment by php-coder** [comment]
No, there shouldn't be my commits, only yours :)

**Comment by php-coder** [comment]
BTW for a new PR -- don't forget to move a file with migration to 0.4.2 folder as 0.4.1 milestone is closed.

**Pull request closed without merging**